### ISSUE TITLE
memory usage continue increase for JointAutoregressiveHierarchicalPriors

### ISSUE THREAD

**Issue by wl3b10s** [InterDigitalInc/CompressAI#335]
when i use JointAutoregressiveHierarchicalPriors to compress and decompress file, the memory usage is increasing along with increase of call times.

and finally the python process is killed by os.

for programs looks like:

for idx in range(count):
    bit = testmodel.compress(test.float())
    dec = testmodel.decompress(bit["strings"], bit["shape"])

I tried to replace the string in _compress_ar by a fixed string and avoid calling BufferedRansEncoder, the memory increase is still exist. So I guess it's not come from the backend implementation of rans.

is there any operation request if i intend to call compress/decompress continuously?

thanks.

**Comment by wl3b10s** [comment]
after debugging by commenting code. 

the following two lines result in memory increase:


_compress_ar -> y_hat[:, :, h + padding[0], w + padding[1]] = y_q + means_hat   
_decompress_ar ->_hat[:, :, hp : hp + 1, wp : wp + 1] = rv 

if I comment out these two lines, the memory will not increase continously.

**Comment by YodaEmbedding** [comment]
Can you print out:

```python
print(f"{y_hat.shape=}")
print(f"{torch.cuda.get_device_properties(0)=}")
```

**Comment by wl3b10s** [comment]
> Can you print out:
> 
> print(f"{y_hat.device=}")
> print(f"{y_hat.shape=}")
> print(f"{torch.cuda.get_device_properties(0)=}")


y_hat.device=device(type='cpu')
y_hat.shape=torch.Size([1, 1, 1, 4128])
torch.cuda.get_device_properties(0)=_CudaDeviceProperties(name='NVIDIA RTX A6000', major=8, minor=6, total_memory=48676MB, multi_processor_count=84)

it is recommended that autoregressive routine is run at cpu, right?

**Comment by YodaEmbedding** [comment]
Print:

```python
    def compress(self, x):
        print(f"{torch.is_grad_enabled()=}")

    def decompress(self, strings, shape):
        print(f"{torch.is_grad_enabled()=}")
```

Does the memory usage still go up if you add `@torch.no_grad()`?

```python
    @torch.no_grad()
    def compress(self, x):
        ...

    @torch.no_grad()
    def decompress(self, strings, shape):
        ...
```

> it is recommended that autoregressive routine is run at cpu, right?

One issue is that minor differences in the floating-point values can lead to large accumulations of error over different computation steps. This is especially problematic for the autoregressive routine because it loops: `decode_symbol, get_new_params, decode_symbol, get_new_params, ...`. The lossless decoder expects the exact same params at every time step as at the encoder. Since the params are derived by applying a prediction model using the latest decoded `y_hat`, sufficiently large errors in the params can cause divergence. (There are certain quantization operations that make this process more resilient, though there's always some small probability of failure proportional to the error.) That means whatever computes the `params` should preferably be as deterministic as possible.

The CPU usually gives more deterministic results than a GPU, which is why it is recommended during the autoregressive encoding/decoding process. However, if the GPU works for you without producing results like https://github.com/InterDigitalInc/CompressAI/issues/235#issuecomment-1740940306 and https://github.com/InterDigitalInc/CompressAI/issues/309, then it's fine to use it.

**Comment by wl3b10s** [comment]
> Print:
> 
>     def compress(self, x):
>         print(f"{torch.is_grad_enabled()=}")
> 
>     def decompress(self, strings, shape):
>         print(f"{torch.is_grad_enabled()=}")
> Does the memory usage still go up if you add `@torch.no_grad()`?
> 
>     @torch.no_grad()
>     def compress(self, x):
>         ...
> 
>     @torch.no_grad()
>     def decompress(self, strings, shape):
>         ...
> > it is recommended that autoregressive routine is run at cpu, right?
> 
> One issue is that minor differences in the floating-point values can lead to large accumulations of error over different computation steps. This is especially problematic for the autoregressive routine because it loops: `decode_symbol, get_new_params, decode_symbol, get_new_params, ...`. The lossless decoder expects the exact same params at every time step as at the encoder. Since the params are derived by applying a prediction model using the latest decoded `y_hat`, sufficiently large errors in the params can cause divergence. (There are certain quantization operations that make this process more resilient, though there's always some small probability of failure proportional to the error.) That means whatever computes the `params` should preferably be as deterministic as possible.
> 
> A CPU usually gives more deterministic results than a GPU, which is why it is recommended during the autoregressive encoding/decoding process. However, if the GPU works for you without producing results like [#235](https://github.com/InterDigitalInc/CompressAI/issues/235) and [#279](https://github.com/InterDigitalInc/CompressAI/issues/279), then it's fine to use it.


with the setting of  @torch.no_grad(). memory increase issue doesn't occur any more.
